File: rings.c

/*
 *  Prototype implementation of the optimal rings
 *  by Tamas Budavari <budavari@pha.jhu.edu>
 *  2012-07-23
 *
 *  NOTES:
 *  - Does only one hemisphere (re-run or mirror)
 *  - Will leave a hole on top to be covered later
 *  - Based on PS1 internal draft and the C# code
 *  - Centers seem to match previous output
 *  - No further testing has beed done
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char* argv[])
{
    double d2r = M_PI / 180;

    // parameter 'a' is the cell size in degrees
    double adeg = 3.955;

    // half of 'a' in radians and its atan
    double halfa = adeg / 2 * d2r;
    double halftheta = atan(halfa);

    // loop init
    double d = 0; // starting Decl. - could change this...
    int ring = 0; // ring ID

    FILE *fp = fopen("rings-c.txt", "w");

    while (d < M_PI / 2 - halftheta)
    {
        double dm = d - halftheta; // eq.5
        if (d == 0) dm = 0; // initial

        int m = (int)ceil(M_PI * cos(dm) / halftheta);  // eq.6
        double ip = 2 * M_PI / m; // eq.7
        double dp = atan(tan(d + halftheta) * cos(ip / 2)); // eq.9


        fprintf(fp, "%d  %d\n", ring, m); // ring & # of cells in that ring
        int i; // dump centers of the cells
        for (i=0; i<m; i++)
        {
            // R.A. can use different phase per ring
            double a = i * ip; // + phase (watch wraparound)
            // print Decl. and R.A. in deg
            fprintf(fp, " \t %d   %25.20f   %25.20f\n", i, d/d2r, a/d2r);
        }

        // advance to next ring
        d = halftheta + dp;
        ring++;
    }

    fclose(fp);
    return 0;
}

File: rings.py

from collections import OrderedDict
import numpy as np

def build_rings(a_deg=3.955):
    d2r = np.pi / 180

    half_a = a_deg / 2 * d2r # half_a in radians
    half_a_theta = np.arctan(half_a)

    ring_data = OrderedDict()

    d = 0
    ring = 0

    while d < np.pi / 2 - half_a_theta:
        dm = d - half_a_theta
        if d==0:
            dm=0

        m = int(np.ceil(np.pi * np.cos(dm) / half_a_theta))
        ip = 2 * np.pi / m
        dp = np.arctan(np.tan(d + half_a_theta) * np.cos(ip / 2))

        ring_data[d/d2r] = []
        for i in range(m):
            a = i * ip
            ring_data[d/d2r].append(a/d2r)

        d = half_a_theta + dp
        ring += 1

    return ring_data

def main_cport():
    d2r = np.pi / 180
    a_deg = 3.955 # cell size in degrees

    half_a = a_deg / 2 * d2r # half_a in radians
    half_a_theta = np.arctan(half_a)

    d = 0
    ring = 0

    f = open("rings-py.txt", "w")
    while d < np.pi / 2 - half_a_theta:
        dm = d - half_a_theta
        if d==0:
            dm=0

        m = int(np.ceil(np.pi * np.cos(dm) / half_a_theta))
        ip = 2 * np.pi / m
        dp = np.arctan(np.tan(d + half_a_theta) * np.cos(ip / 2))

        f.write(f"{ring} {m}\n")

        for i in range(m):
            a = i * ip
            f.write(f"{i} {d / d2r} {a / d2r}\n")

        d = half_a_theta + dp
        ring += 1

    f.close()


def compare():
    with open("rings-py.txt", "r") as f, open("rings-c.txt", "r") as g:
        py_lines = [list(map(float, l.strip().split())) for l in f.readlines()]
        c_lines = [list(map(float, l.strip().split())) for l in g.readlines()]

    for i in range(len(py_lines)):
        np.testing.assert_allclose(py_lines[i], c_lines[i])


if __name__=="__main__":
    main_cport()
    compare()
